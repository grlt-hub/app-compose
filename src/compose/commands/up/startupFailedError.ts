import type { ContainerId } from '@createContainer';
import type { StageId } from '@prepareStages';
import { LIBRARY_NAME } from '@shared';
import dedent from 'dedent';

type Params = {
  containerId: ContainerId;
  stageId: StageId;
  log: Record<string, Record<string, unknown>>;
};

const throwStartupFailedError = ({ containerId, stageId, log }: Params) => {
  throw new Error(
    dedent`${LIBRARY_NAME} Application startup failed.
    Required container "${containerId}" did not up in stage "${stageId}".

    Startup Log:
    ${JSON.stringify(JSON.stringify(log, null, 2))
      .replace(/\{/g, '')
      .replace(/\\"/g, '')
      .replace(/[,"]/g, '')
      .replace(/\}\\n  \}/g, '')
      .replace(/\\n\}/g, '')}

    Recommendations:
      - Verify if "${containerId}" is truly required.
      - If not, consider removing it from the required list in "up.required".
      - Check that all its dependencies are correct and its logic works as expected.
  `,
  );
};

export { throwStartupFailedError };
