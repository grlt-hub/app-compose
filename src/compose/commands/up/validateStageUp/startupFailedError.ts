import type { ContainerId } from '@createContainer';
import type { StageId } from '@prepareStages';
import { LIBRARY_NAME } from '@shared';

type Params = {
  id: ContainerId | ContainerId[];
  stageId: StageId;
  log: Record<string, Record<string, unknown>>;
};

const throwStartupFailedError = ({ id, stageId, log }: Params) => {
  throw new Error(
    `${LIBRARY_NAME} Application startup failed.` +
      '\n' +
      `Required container(s) "${id}" did not up in stage "${stageId}".` +
      '\n\n' +
      'Startup Log:' +
      '\n' +
      `${JSON.stringify(log, null, 2)
        .replace(/[\{\},"]/g, '')
        .replace(/\n\s*\n/g, '\n\n')}` +
      '\n' +
      'Recommendations:' +
      '\n' +
      `- Verify if the container(s) "${id}" are truly required.` +
      '\n' +
      `- If not, consider removing them from the required list in "up.required".` +
      '\n' +
      `- Ensure all dependencies for the container(s) are correct and their logic works as expected.`,
  );
};

export { throwStartupFailedError };
