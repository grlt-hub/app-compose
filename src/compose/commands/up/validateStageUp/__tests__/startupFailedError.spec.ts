import { describe, expect, it } from 'vitest';
import { throwStartupFailedError } from '../startupFailedError';

describe('throwStartupFailedError', () => {
  it('should throw an error with the correct message', () => {
    const id = 'test-container';
    const stageId = 'test-stage';
    const log = {
      'entities-stage': {
        allSucceeded: true,
        containerStatuses: {
          entities: 'done',
        },
      },
      'notifications-stage': {
        allSucceeded: false,
        containerStatuses: {
          notifications: 'fail',
        },
      },
    };

    expect(() => throwStartupFailedError({ id, stageId, log })).toThrowErrorMatchingSnapshot();
  });

  it('should handle multiple container IDs correctly', () => {
    const id = ['container1', 'container2'];
    const stageId = 'another-stage';
    const log = {
      'entities-stage': {
        allSucceeded: false,
        containerStatuses: {
          container1: 'idle',
          container2: 'pending',
        },
      },
    };

    expect(() => throwStartupFailedError({ id, stageId, log })).toThrowErrorMatchingSnapshot();
  });
});
