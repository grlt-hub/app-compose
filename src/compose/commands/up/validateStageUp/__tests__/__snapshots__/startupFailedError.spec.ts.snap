// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`throwStartupFailedError > should handle multiple container IDs correctly 1`] = `
[Error: [app-compose] Application startup failed.
Required container(s) "container1,container2" did not up in stage "another-stage".

Startup Log:

  entities-stage: 
    allSucceeded: false
    containerStatuses: 
      container1: idle
      container2: pending


Recommendations:
- Verify if the container(s) "container1,container2" are truly required.
- If not, consider removing them from the required list in "up.required".
- Ensure all dependencies for the container(s) are correct and their logic works as expected.]
`;

exports[`throwStartupFailedError > should throw an error with the correct message 1`] = `
[Error: [app-compose] Application startup failed.
Required container(s) "test-container" did not up in stage "test-stage".

Startup Log:

  entities-stage: 
    allSucceeded: true
    containerStatuses: 
      entities: done

  notifications-stage: 
    allSucceeded: false
    containerStatuses: 
      notifications: fail


Recommendations:
- Verify if the container(s) "test-container" are truly required.
- If not, consider removing them from the required list in "up.required".
- Ensure all dependencies for the container(s) are correct and their logic works as expected.]
`;
